File: JavaPatterns/src/main/java/creational/java7/Factory.java
package creational.java7;

/**
 * The factory method pattern provides an interface for creating an object that encapsulates the actual class instantiation in a method,
 * and lets subclasses decide which class to instantiate.
 */
interface Animal {
    void speak();
}

class AnimalFactory {

    private static AnimalFactory instance = null;

    private AnimalFactory() {
    }

    public static AnimalFactory getInstance() {
        if (instance == null) {
            instance = new AnimalFactory();
        }
        return instance;
    }

    private class Dog implements Animal {
        @Override
        public void speak() {
            System.out.println("Hello, I'm a Java dog!");
        }
    }

    private class Cat implements Animal {
        @Override
        public void speak() {
            System.out.println("Java Miau");
        }
    }


    public Animal createAnimal(String kind) {
        if ("cat".equals(kind)) return new Cat();
        if ("dog".equals(kind)) return new Dog();
        throw new IllegalArgumentException();
    }

}
public class Factory {

    public static void main(String[] args) {
        Animal dog = AnimalFactory.getInstance().createAnimal("dog");
        Animal cat = AnimalFactory.getInstance().createAnimal("cat");

        dog.speak();
        cat.speak();

    }
}

File: ScalaPatterns/src/main/scala/creational/Factory.scala
package creational

/**
 * The factory method pattern provides an interface for creating an object that encapsulates the actual class instantiation in a method,
 * and lets subclasses decide which class to instantiate.
 *
 * Scala provides a special syntactic construct, that looks similar to constructor invocation, but it’s actually a convenient factory method
 *
 * @author Daniel Leon
 */
object Factory {

  trait Animal {
    def speak
    def name
  }

  private class Dog extends Animal {
    override def speak {
      println("Hello, I'm a Scala dog!")
    }
    override def name {
      "Dog"
    }
  }

  private class Cat extends Animal {
    override def speak {
      println("Scala Meow")
    }

    override def name {
      "Cat"
    }
  }

  object Animal {
    def apply(kind: String) = kind match {
      case "dog" => new Dog()
      case "cat" => new Cat()
    }
  }

  def main(args: Array[String]) {
    val dog = Animal("dog")
    val cat = Animal("cat")

    dog.speak
    cat.speak
  }
}
